ihatemoney/web.py

```diff
@@ -195,15 +195,18 @@ def admin():
 def authenticate(project_id=None):
     """Authentication form"""
     form = AuthenticationForm()
+
+    if not form.id.data and request.args.get("project_id"):
+        form.id.data = request.args["project_id"]
+    project_id = form.id.data
     # Try to get project_id from token first
     token = request.args.get("token")
     if token:
-        project_id = Project.verify_token(token, token_type="non_timed_token")
+        project_id = Project.verify_token(
+            token, token_type="auth", project_id=project_id
+        )
```

Comment:
Interesting code :)

Just to be safe, I would check that the project ID of the payload is equal to the project ID in the GET parameter. There could be weird cases where they are different. For instance, what happens if projects A and B have the same private code and I give the token of A to authenticate against B?

Reply:
This is done in verify_token:
```python
return (
    data_project if project_id is None or data_project == project_id else None
)
```

This is tested in `test_invite()` when I change the `project_id` of a valid token.

Reply:
Ok, but defensive programming is good: `verify_token` might change in the future.

Either we should ignore the return value, or else we should check it's consistent. But overwriting project_id like that is a source of future bugs.

Reply:
Done in bb5c7abd045b10828ff70cda1fe5a3210e73279d.